### ISSUE TITLE
Disable force push

### ISSUE THREAD

**Issue by okuryu** [mesosphere/marathon#2198]
I'm trying to build Marathon [v0.9.0](https://github.com/mesosphere/marathon/tree/v0.9.0), but Marathon v0.9.0 source tree fails to resolve Git submodules (marathon-ui) because it has lost a commit to build Marathon. This issue occurs on Marathon v0.9.0, [v0.9.1](https://github.com/mesosphere/marathon/tree/v0.9.1), and [v0.9.2](https://github.com/mesosphere/marathon/tree/v0.9.2).

Checkout the marathon and update Git submodules.

```
$ git checkout v0.9.0
$ git submodule update --init --recursive
Submodule 'src/main/resources/assets' (https://github.com/mesosphere/marathon-ui.git) registered for path 'src/main/resources/assets'
Cloning into 'src/main/resources/assets'...
remote: Counting objects: 9097, done.
remote: Compressing objects: 100% (223/223), done.
remote: Total 9097 (delta 139), reused 0 (delta 0), pack-reused 8872
Receiving objects: 100% (9097/9097), 6.67 MiB | 2.04 MiB/s, done.
Resolving deltas: 100% (6482/6482), done.
Checking connectivity... done.
fatal: reference is not a tree: 67bc3499f2a3bcf58008f1ca71747dcf72e44941
Unable to checkout '67bc3499f2a3bcf58008f1ca71747dcf72e44941' in submodule path 'src/main/resources/assets'
```

Check Git submodule.

```
$ git submodule status
-67bc3499f2a3bcf58008f1ca71747dcf72e44941 src/main/resources/assets
```

A commit on marathon-ui is `67bc3499f2a3bcf58008f1ca71747dcf72e44941`. You can look this commit on GitHub.

* https://github.com/mesosphere/marathon-ui/commit/67bc3499f2a3bcf58008f1ca71747dcf72e44941

However, this commit actually doesn't exist anywhere.

```
$ git clone https://github.com/mesosphere/marathon-ui.git
Cloning into 'marathon-ui'...
remote: Counting objects: 9097, done.
remote: Compressing objects: 100% (223/223), done.
remote: Total 9097 (delta 139), reused 0 (delta 0), pack-reused 8872
Receiving objects: 100% (9097/9097), 6.67 MiB | 2.04 MiB/s, done.
Resolving deltas: 100% (6482/6482), done.
Checking connectivity... done.
$ cd marathon-ui
$ git show 67bc3499f2a3bcf58008f1ca71747dcf72e44941
fatal: bad object 67bc3499f2a3bcf58008f1ca71747dcf72e44941
```

Also, Marathon v0.9.2 depends on marathon-ui commit `f296698a033df99539c8bbd75ebfcfb1b66eb8a6`, the same issue occurs.

* https://github.com/mesosphere/marathon-ui/commit/f296698a033df99539c8bbd75ebfcfb1b66eb8a6

```
$ git show f296698a033df99539c8bbd75ebfcfb1b66eb8a6
fatal: bad object f296698a033df99539c8bbd75ebfcfb1b66eb8a6
```

My guess that this issue is a change of Git history from any cause like performing `git push -f`. This means there is no way to build Marathon v0.9.x right now.

I hope that turn on [disable force push](https://github.com/blog/2051-protected-branches-and-required-status-checks) to marathon and marathon-ui repositories to prevent a recurrence this issue. And I appreciate you provide a patch release like v0.9.3 in fixing this issue if possible.

**Comment by philipnrmn** [comment]
Hi @okuryu,

Thanks very much for your detailed bug report. 

First the good news: this issue is now fixed and tested on all 0.9.x versions. 

We already have a policy of never force-pushing on `master`, but it doesn't hurt to enforce it, so we will take your recommendation. The feature is still being rolled out and hasn't reached `marathon-ui` yet, but when it does we'll be sure to take advantage of it 

In this case however it wasn't actually force-push that caused the bad reference. The commits which the three releases referenced between them were never merged into master. Their branches were recently deleted during a post-tag spring-cleaning session. 

We were able to restore the branches on which they lived and will ensure that they are not deleted again. We will also attempt to find a way of non-destructively merging them into master so that they are safe in the future.
